src/libraries/Common/src/SourceGenerators/DiagnosticInfo.cs

```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Linq;
+using System.Numerics.Hashing;
+using Microsoft.CodeAnalysis;
+
+namespace SourceGenerators;
+
+/// <summary>
+/// Descriptor for diagnostic instances using structural equality comparison.
+/// Provides a work-around for https://github.com/dotnet/roslyn/issues/68291.
+/// </summary>
+internal readonly struct DiagnosticInfo : IEquatable<DiagnosticInfo>
+{
+    public required DiagnosticDescriptor Descriptor { get; init; }
+    public required object?[] MessageArgs { get; init; }
+    public required Location? Location { get; init; }
```

Comment:
After discussion with @layomia, I missed a call to `GetTrimmedLocation()` that would address this. I'd suggest making that call in this type, rather than where it's created, so we can never miss it.

Reply:
Done; let's consider this addressed for this PR. Created https://github.com/dotnet/runtime/issues/92509 to follow up.